modules/features/search/src/main/java/au/com/shiftyjelly/pocketcasts/search/SearchViewModel.kt

```diff
@@ -51,43 +46,50 @@ class SearchViewModel @Inject constructor(
+                                    analyticsTracker.track(
+                                        AnalyticsEvent.IMPROVED_SEARCH_SUGGESTIONS_FAILED,
+                                        mapOf(
+                                            "source" to source,
+                                        ),
+                                    )
                                 }
 
-                                // only show loading for the initial query when autocomplete results are empty
-                                if (((it as? SearchUiState.Suggestions)?.operation as? SearchUiState.SearchOperation.Success)?.results?.isNotEmpty() == true && operation is SearchUiState.SearchOperation.Loading) {
-                                    it
-                                } else {
-                                    SearchUiState.Suggestions(operation = operation)
-                                }
+                                else -> Unit
+                            }
+
+                            // only show loading for the initial query when autocomplete results are empty
+                            if (((uiState as? SearchUiState.Suggestions)?.operation as? SearchUiState.SearchOperation.Success)?.results?.isNotEmpty() == true && operation is SearchUiState.SearchOperation.Loading) {
+                                uiState
                             } else {
-                                it
+                                SearchUiState.Suggestions(operation = operation)
                             }
                         }
                     }
             }
         }
 
         viewModelScope.launch {
-            searchHandler.searchResults.collect {
-                if (_state.value is SearchUiState.Results || !FeatureFlag.isEnabled(Feature.IMPROVED_SEARCH_SUGGESTIONS)) {
-                    _state.value = SearchUiState.Results(operation = it as SearchUiState.SearchOperation<SearchResults>)
+            if (FeatureFlag.isEnabled(Feature.IMPROVED_SEARCH_RESULTS)) {
+                searchHandler.improvedSearchResults.collect {
+                    showSearchHistory = false
+                    _state.value = SearchUiState.ImprovedResults(operation = it)
+                    if (!FeatureFlag.isEnabled(Feature.IMPROVED_SEARCH_SUGGESTIONS) && it is SearchUiState.SearchOperation.Loading) {
+                        saveSearchTerm(it.searchTerm)
+                    }
                 }
-
-                if (!FeatureFlag.isEnabled(Feature.IMPROVED_SEARCH_SUGGESTIONS) && it is SearchUiState.SearchOperation.Loading) {
-                    saveSearchTerm(it.searchTerm)
+            } else {
+                searchHandler.searchResults.collect {
                     showSearchHistory = false
+                    if (_state.value is SearchUiState.OldResults) {
+                        _state.value = SearchUiState.OldResults(operation = it as SearchUiState.SearchOperation<SearchResults.SegregatedResults>)
```

Comment:
❓ How do we know that the `it as SearchUiState.SearchOperation<SearchResults.SegregatedResults` casting won't fail?